rfcs/0000-local-overlay-store.md

```diff
@@ -0,0 +1,408 @@
+
+`local-overlay` is a store representing the extension of a lower store with a collection of additional store objects.
+(We don't refer to the upper layer as an "upper store" because it is not self-contained
+--- it doesn't abide by the closure property because objects in the upper layer can refer to objects that are only in the lower layer.)
+
+## Class hierarchy, configuration settings, and initialization
+
+```mermaid
+flowchart TD
+    subgraph LocalFSStore-instance
+        LS
+        LM
+    end
+
+    LS(store directory)
+    LM(Abstract metadata source)
+
+    subgraph LocalOverlayStore-instance
+        US
+        UM
+    end
+
+    US(store directory)
+    UM(SQLite DB)
+
+    UD(Directory for additional store objects)
+
+    LocalOverlayStore-instance -->|`lower-store` configuation option| LocalFSStore-instance
+    LocalOverlayStore-instance -->|`upper-layer` configuation option| UD
+
+    US -->|OverlayFS lower layer| LS
+    US -->|OverlayFS upper layer| UD
+```
+
+`LocalOverlayStore` is a subclass of `LocalStore` implementing the `local-overlay` store.
+It has additional configuration items for:
+
+ - `lower-store`: The lower store, which must be a `LocalFSStore`
+
+   This is specified with an escaped URL just like the `remote-store` setting of the two SSH stores types.
+
+ - `upper-layer`: The directory used as the upper layer of the OverlayFS
+
+ - `check-mount`: Whether to check the filesystem mount configuration
+
+   Here is an example of how the nix.conf configuration might look:
+   ```
+   store = local-overlay?lower-store=/mnt/lower/%3Fread-only%3Dtrue&upper-layer=/mnt/scratch/nix/upper/store
+   ```
+
+With `check-mount` enabled, on initialization it checks that an OverlayFS mount exists matching these parameters:
+
+ - The lower layer must be the lower store's "real store directory"
+
+ - The upper layer must be the directory specified for this purpose
+
+The database for the `local-overlay` store is exactly like that for a regular local store:
+
+ - Same schema, including foreign key constraints
+
+ - Created empty on opening the store if it doesn't exist
+
+ - Opened existing one otherwise
+
+## Data structure
+
+These are diagrams for the possible states of a single store object.
+Except for the closure property mandating that present store objects must have all their references also be present, store objects are independent.
+We can therefore to a large extent get away with considering only a single store object.
+
+### Graph Key
+
+All the graphs below follow these conventions:
+
+#### Nodes
+
+- Right angle corners: Absent store object
+- Rounded corners: Present store object
+
+#### Edges
+
+- Solid edge: "adding" or "deduplicating" direction
+- Dotted edge: "deleting" direction
+
+The graph when reduced to just one type of edge is always acyclic, and thus represents a partial order.
+
+### Lower Store
+
+While in use as a lower store by one or more overlay stores, a store most only grow "monotonically".
```

Comment:
In Linux, [changes to the underlying storage of overlay devices has undefined behaviour](https://www.kernel.org/doc/html/latest/filesystems/overlayfs.html#changes-to-underlying-filesystems), so a store used as a lower store cannot grow at all.